src/dlangui/graphics/colors.d

```diff
@@ -312,44 +312,100 @@ bool isFullyTransparentColor(uint color) pure nothrow {
+        if (r < 0 || r > 255 || g < 0 || g > 255 || b < 0 || b > 255)
+            return defValue;
+        uint a = 255;
+        auto ap = parts[3].strip;
+        if (ap.endsWith("%")) { //rgba(r,g,b,a%)
+            auto alpha = to!float(ap[0 .. $ - 1].strip);
+            if(alpha < 0 || alpha > 100)
+                return defValue;
+            a = cast(uint)((alpha * 255.0 / 100.0) + 0.5);
+        }
+        else { //rgba(r,g,b,a)
+            auto alpha = to!float(parts[3].strip);
+            if (alpha < 0 || alpha > 1)
+                return defValue;
+            a = cast(uint)(alpha * 255.0 + 0.5);
+        }
+        return (a << 24) | (r << 16) | (g << 8) | b;
+    }
+    else if (s.startsWith("rgb(") && s.endsWith(")"))
+    {
+        s = s[4 .. $ - 1];
+        auto parts = s.split(",");
+        if (parts.length != 3)
+            return defValue;
+        uint r = to!uint(parts[0].strip);
+        uint g = to!uint(parts[1].strip);
+        uint b = to!uint(parts[2].strip);
+        if (r < 0 || r > 255 || g < 0 || g > 255 || b < 0 || b > 255)
             return defValue;
-        value = (value << 4) | digit;
-        if (s.length < 7) // double the same digit for short forms
-            value = (value << 4) | digit;
+        return (r << 16) | (g << 8) | b;
+    }
+    foreach (color; __traits(allMembers, Color))
+    {
+        if (color == s)
+            return color.to!Color;
     }
-    return value;
+    return defValue;
+}
+
+unittest
+{
+    static assert(decodeHexColor("") == 0);
+    static assert(decodeHexColor("@null") == COLOR_TRANSPARENT);
+    static assert(decodeHexColor("trAnsParent") == COLOR_TRANSPARENT);
+    static assert(decodeHexColor("grAy") == 0x808080);
+    static assert(decodeHexColor("#8B008B") == 0x8B008B);
+    static assert(decodeHexColor("#fFf") == 0xfff);
+    static assert(decodeHexColor("#f0F0") == 0xf0f0);
+    static assert(decodeHexColor("#80ff0000") == 0x80ff0000);
+    static assert(decodeHexColor("rgba(255, 0, 0,.5 )") == 0x80ff0000);
+    static assert(decodeHexColor("rgba(255,0, 0, 50%)") == 0x80ff0000);
+    static assert(decodeHexColor("rgba(255,0, 0, 100%)") == 0xffff0000);
+    static assert(decodeHexColor("rgba(255,0, 0, 0%)") == 0xff0000);
+    static assert(decodeHexColor("rgb(255,255, 255)") == 0xffffff);
+    static assert(decodeHexColor("rgba(255,0, 0, 150%)") == 0); // invalid input, return def value
+    static assert(decodeHexColor("rgba(255,0, 0, -34%)") == 0); // invalid input, return def value
+    static assert(decodeHexColor("rgb(321,321,321)") == 0); // invalid input, return def value
+    static assert(decodeHexColor("not_valid_color_name") == 0); // invalid input, return def value
+    static assert(decodeHexColor("#80ff00000") == 0); // invalid input, return def value
+    static assert(decodeHexColor("#f0") == 0); // invalid input, return def value
+    static assert(decodeHexColor("rgba(255,255, 255, 10)") == 0); // invalid input, return def value
```

Comment:
Shouldn't this be valid though? As in here it can reinterpretate 10 as being an ubyte and just do 10/255. I don't really know how CSS works though

Reply:
In CSS, the value is cast to the nearest valid value in the range 0-1. That is, in the case of rgba(255,0,0,10) we get rgba(255,0,0,1). Should I do the same?